### PR TITLE
prompt templating

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

**Comment by sjwhitmore** [comment]
this looks great! very much in line with what i was thinking

File: langchain/prompt.py [hunk 1]
```diff
@@ -70,3 +70,19 @@ def template_is_valid(cls, values: Dict) -> Dict:
...
+    def from_examples(
+        cls,
+        examples: List[Dict],
+        suffix: str,
+        input_variables: List[str],
+        example_prompt: "Prompt",
+        example_separator: str = "\n",
+        prefix="",
+    ):
+        example_str = example_separator.join(
```
Comment (sjwhitmore) [inline langchain/prompt.py:84]: could add something along the lines of
```for e in examples, assert(input_variables in e.keys())... ```
and something along the lines of checking the suffix is also aligned with one of the keys in the examples.
just to make sure the formatting of the examples match the formatting of the input variables and output type for the prompt.

**Comment by sjwhitmore** [comment]
After examining the structure of the chained prompts (e.g. self ask with search & ReAct & Natbot), I think a less opinionated approach to the structure of the examples (e.g. a list of examples as strings rather than a list of examples as dicts with specific key-value pairs) is more likely to generalize across all prompts. I think the value of a list of examples is 1) readability 2) dynamically omitting an example if my input to the prompt is too long. I don't think we need to be too opinionated about the key/value pairs in the examples as we're already trusting the user to structure their prompt templates correctly... we don't want to make it too onerous to fit every example into the key/value input_variables + output_suffix paradigm. I'll push an update which simply accepts examples as a list.

**Comment by sjwhitmore** [comment]
Pushed an example with the ReAct template... not quite sure how best to work in the example sizing into the Chain class.
Option 1: Add a num_examples argument to the __call__() method which takes a splice of the list of examples and uses that to initialize prompt in the _run() method. This feels like it muddies up the interface a bit, and also requires knowledge of the examples themselves and how long they are.
Option 2: Add a dynamic_prompt_sizing=True argument to the config, which detects if a prompt edges towards the token limit (or receives an over token limit error from an LLM) and automatically cuts an example. This feels like it might be a bit too magical / hiding too much config from the user.
Option 3: Explicitly let the user choose to use the example splicing in any further classes they define, and don't integrate it into the classes here.
Option 4: ?? Am I missing anything here?

**Comment by hwchase17** [comment]
yeah i like the change to go to whole examples. we can also do both! expose different constructor methods depending on the use case

related to your last point around incorporating dynamic examples, what do you think about just having a different prompt class? this prompt class could be like `DynamicPrompt`, and could take in init a list of examples, and then the dynamic removal of examples (if necc) comes into play when you call `.format`?

**Comment by hwchase17** [comment]
potentially theres two separate ideas/prs here:

- make it easy to specify a prompt from examples (what currently is here)
- make a separate dynamic prompt class (in follow up pr)

**Comment by sjwhitmore** [comment]
makes sense! I like the idea of adding a DynamicPrompt in a separate PR.
I forgot to add the 3rd reason for having an explicit list of examples (which originally motivated my comment about this over text haha) -- making it easier to generate test data more easily.

File: langchain/chains/react/prompt.py [hunk 2]
```diff
@@ -109,9 +104,12 @@
 scientist.
 Thought 3: Leonid Levin is a mathematician and computer scientist. So Pavel Urysohn
 and Leonid Levin have the same type of work.
-Action 3: Finish[yes]
+Action 3: Finish[yes]"""]
+PREFIX=""
+EXAMPLE_SEPARATOR = """\n\n"""
+SUFFIX = """\n\nQuestion: {input}"""
+_DEFAULT_TEMPLATE = PREFIX + EXAMPLE_SEPARATOR.join(EXAMPLES) + SUFFIX
```
Comment (hwchase17) [inline langchain/chains/react/prompt.py:111]: can we use the `from_examples` classmethod here?

File: langchain/prompt.py [hunk 3]
```diff
@@ -70,3 +73,17 @@ def template_is_valid(cls, values: Dict) -> Dict:
...
+
+    @classmethod
+    def from_examples(
+        cls: Type[TPrompt],
+        examples: List[str],
+        suffix: str,
+        input_variables: List[str],
+        example_separator: str = "\n",
+        prefix: str = "",
+    ) -> TPrompt:
```
Comment (hwchase17) [inline langchain/prompt.py:85]: i think you can just (1) not type hint `cls` and have the return type as `"Prompt"` (like in quotes) and you should be good

**Review by sjwhitmore — APPROVED** [review]
lgtm!

**Pull request merged**